core/src/main/scala/kafka/utils/ShutdownableThread.scala

```diff
@@ -75,7 +81,12 @@ abstract class ShutdownableThread(val name: String, val isInterruptible: Boolean
    */
   def doWork(): Unit
 
+  override def start(): Unit = {
+    isStarted = true
+    super.start()
+  }
   override def run(): Unit = {
+    isStarted = true
```

Comment:
Hmm, do we need to set isStarted here again since it's already set in start()?

Reply:
I was trying to cover the case where `run()` is called directly, without `start()`, either directly by a caller of by an executor implementation.

Reply:
In that case, could we just set it in run() and not in start()?

Reply:
I added in the `start()` case because there could be a delay between `start()` getting called and `run()` getting called in the new thread, but I think this sends up producing inconsistent behavior depending on how the `ShutdownableThread` is started.

In the current state, if `start()` is used to start the `ShutdownableThread`, we're guaranteed to always correctly block in subsequent calls to `awaitShutdown()` since `isStarted` is set from the caller thread. If the `ShutdownableThread` is submitted to an executor, `run()` can be called at any time in the future, which makes it impossible for the caller thread to know if `awaitShutdown()` will block as we expect. 

We end up with different guarantees based on how the class is run, which is not optimal. I think the only one we can guarantee across both `start()` and an Executor starting the thread is that the caller thread won't know when `isStarted` is true, and as a result, when it can rely on the blocking behavior of `awaitShutdown`.